File: app/src/main/java/com/asvorded/monidroid/EchoClient.java
package com.asvorded.monidroid;

import android.util.Log;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Objects;
import java.util.function.Consumer;

public class EchoClient {

    public static class HostInfo {
        private final InetAddress address;
        private final String hostName;

        public HostInfo(InetAddress address, String hostName) {
            this.address = address;
            this.hostName = hostName;
        }

        public String getHostName() {
            return hostName;
        }

        public InetAddress getAddress() {
            return address;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            HostInfo hostInfo = (HostInfo) o;
            return Objects.equals(address, hostInfo.address) && Objects.equals(hostName, hostInfo.hostName);
        }

        @Override
        public int hashCode() {
            return Objects.hash(address, hostName);
        }
    }

    public enum AutoDetectingOptions {
        Enabled, Error, Disabled
    }

    public EchoClient() {
        echoSendThread = new Thread(this::sendMain);
        echoReceiveThread = new Thread(this::receiveMain);
    }

    private final Thread echoSendThread;
    private final Thread echoReceiveThread;

    private DatagramSocket echoSocket;

    private Runnable onFailed;
    private Consumer<HostInfo> onDeviceDetected;

    public void startEcho(Runnable failCallback, Consumer<HostInfo> deviceDetectedCallback) throws IOException {
        echoSocket = new DatagramSocket(MonidroidProtocol.MONITOR_PORT);
        echoSocket.setBroadcast(true);

        onDeviceDetected = deviceDetectedCallback;
        onFailed = failCallback;

        echoSendThread.start();
        echoReceiveThread.start();
    }

    public void startEcho2(Runnable failCallback, Consumer<HostInfo> deviceDetectedCallback) throws IOException {
        echoSocket = new DatagramSocket(MonidroidProtocol.MONITOR_PORT);
        echoSocket.setBroadcast(true);

        onDeviceDetected = deviceDetectedCallback;
        onFailed = failCallback;

        echoReceiveThread.start();
    }

    private void sendMain() {
        byte[] sendBuf = MonidroidProtocol.CLIENT_ECHO_WORD.getBytes(StandardCharsets.US_ASCII);
        InetAddress broadcastAddress;
        try {
            broadcastAddress = InetAddress.getByName("255.255.255.255");
        } catch (IOException ignored) {
            return;
        }
        DatagramPacket dgram = new DatagramPacket(
            sendBuf, sendBuf.length, broadcastAddress, MonidroidProtocol.MONITOR_PORT
        );

        boolean sending = true;
        while (sending) {
            try {
                onDeviceDetected.accept(null);

                echoSocket.send(dgram);
                Log.d(MonidroidProtocol.DEBUG_TAG, "Successfully sent datagram, waiting 5 seconds");
                Thread.sleep(5000);
            } catch (InterruptedException e) {
                return;
            } catch (IOException e) {
                sending = false;
                onFailed.run();
            }
        }
    }

    private void receiveMain() {
        byte[] header = MonidroidProtocol.SERVER_ECHO_WORD.getBytes(StandardCharsets.US_ASCII);
        byte[] buf = new byte[128];

        byte[] headerBuf = new byte[header.length];
        byte[] lenBuf = new byte[4];

        boolean accepting = true;
        while (accepting) {
            try {
                DatagramPacket dgram = new DatagramPacket(buf, buf.length);
                echoSocket.receive(dgram);
                if (dgram.getLength() > header.length + 4) {
                    byte[] data = dgram.getData();

                    // Check header
                    System.arraycopy(data, 0, headerBuf, 0, header.length);
                    if (Arrays.equals(header, headerBuf)) {
                        // Get host name length
                        System.arraycopy(data, header.length, lenBuf, 0, 4);

                        int len = ByteBuffer.wrap(lenBuf)
                                .order(ByteOrder.LITTLE_ENDIAN)
                                .getInt();

                        if (header.length + 4 + len * 2 == dgram.getLength()) {
                            // Get host name
                            byte[] nameBuf = new byte[len * 2];
                            System.arraycopy(data, header.length + 4, nameBuf, 0, nameBuf.length);
                            String hostName = new String(nameBuf, StandardCharsets.UTF_16LE);

                            // Trigger event
                            HostInfo hostInfo = new HostInfo(dgram.getAddress(), hostName);
                            onDeviceDetected.accept(hostInfo);
                        }
                    }
                }
            } catch (IOException e) {
                accepting = false;
                onFailed.run();
            }
        }
    }

    public void endEcho() {
        echoSendThread.interrupt();
        echoReceiveThread.interrupt();
        echoSocket.close();
    }

    public void endEcho2() {
        echoSendThread.interrupt();
        echoSocket.close();
    }
}

File: app/src/main/java/com/asvorded/monidroid/EchoClient.kt
package com.asvorded.monidroid

import com.asvorded.monidroid.MonidroidProtocol.SV_ECHO_WORD
import java.io.IOException
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.NetworkInterface
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets
import java.util.Collections

class EchoClient {
    data class HostInfo(val address: InetAddress, val hostName: String?)

    enum class AutoDetectingOptions {
        Enabled, Error, Disabled
    }

    var started = false
        private set

    private val foundHosts = Collections.synchronizedSet(mutableSetOf<HostInfo>())

    private lateinit var echoSendThread: Thread
    private lateinit var echoReceiveThread: Thread

    private lateinit var echoSocket: DatagramSocket

    var onFailed: (() -> Unit)? = null
    var onDevicesDetected: ((Set<HostInfo>) -> Unit)? = null

    fun start() {
        resume()
    }

    fun pause() {
        stop()
    }

    fun resume() {
        if (started) return

        echoSocket = DatagramSocket(MonidroidProtocol.PROTOCOL_PORT)
        echoSocket.broadcast = true

        echoSendThread = Thread { sendMain() }
        echoSendThread.start()

        echoReceiveThread = Thread { receiveMain() }
        echoReceiveThread.start()

        started = true
    }


    fun stop() {
        if (!started) return

        started = false
        // now threads see updated `started`
        echoSendThread.interrupt()
        echoReceiveThread.interrupt()
        echoSocket.close()
    }


    private fun sendMain() {
        val sendBuf = MonidroidProtocol.ECHO_WORD.toByteArray(StandardCharsets.US_ASCII)

        while (true) {
            val saved = foundHosts.toSet()
            foundHosts.clear()

            try {
                // send broadcast to all interfaces
                for (ni in NetworkInterface.getNetworkInterfaces()) {
                    if (!ni.isUp || ni.isLoopback) continue
                    for (addr in ni.interfaceAddresses) {
                        val broadcast = addr.broadcast
                        if (broadcast != null) {
                            val dgram = DatagramPacket(
                                sendBuf, sendBuf.size, broadcast,
                                MonidroidProtocol.PROTOCOL_PORT
                            )
                            echoSocket.send(dgram)
                        }
                    }
                }
                Thread.sleep(1000)

                // diff saved state and current
                onDevicesDetected?.invoke(foundHosts intersect saved)
            } catch (_: InterruptedException) {
                return
            } catch (_: IOException) {
                if (started) onFailed?.invoke()
                break
            }
        }
    }

    private fun receiveMain() {
        val buf = ByteArray(128)

        var accepting = true
        while (accepting) {
            try {
                val dgram = DatagramPacket(buf, buf.size)
                echoSocket.receive(dgram)
                if (dgram.address != echoSocket.inetAddress && dgram.length >= SV_ECHO_WORD.length + 4) {
                    val data = dgram.data

                    // Check header
                    if (data.copyOfRange(0, SV_ECHO_WORD.length)
                        .toString(StandardCharsets.UTF_8) == SV_ECHO_WORD)
                    {
                        // Get host name length
                        val len = ByteBuffer
                            .wrap(data.copyOfRange(SV_ECHO_WORD.length, SV_ECHO_WORD.length + 4))
                            .order(ByteOrder.LITTLE_ENDIAN)
                            .getInt()

                        if (SV_ECHO_WORD.length + 4 + len == dgram.length) {
                            val hostName = data.copyOfRange(SV_ECHO_WORD.length + 4, dgram.length)
                                .toString(StandardCharsets.UTF_8)

                            foundHosts += HostInfo(dgram.address, hostName)
                        }
                    }
                }
            } catch (_: IOException) {
                accepting = false
                if (started) onFailed?.invoke()
            }
        }
    }
}